File: apps/api/src/sync/worksheet-mapper.ts
export interface RowMappingConfig {
  dateCol: number;
  rollNoCol: number;
  statusCol: number;
  facultyCol?: number;
  remarksCol?: number;
}

export interface ParsedAttendanceRow {
  studentRollNo: string;
  lectureDate: Date;
  status: 'PRESENT' | 'ABSENT' | 'NA';
  faculty?: string;
  remarks?: string;
  sourceRowHash: string;
}

export class WorksheetMapper {
  /**
   * Coerce roll number into clean uppercase format (e.g., '2024cs01' -> '2024CS01')
   */
  static coerceRollNo(raw?: string): string | null {
    if (!raw) return null;
    const clean = String(raw).trim().toUpperCase();
    return clean.length > 0 ? clean : null;
  }

  /**
   * Coerce date string or timestamp into valid Date object
   */
  static coerceDate(raw?: string): Date | null {
    if (!raw) return null;
    const cleanStr = String(raw).trim();
    if (!cleanStr) return null;

    const parsed = new Date(cleanStr);
    if (isNaN(parsed.getTime())) {
      return null;
    }
    return parsed;
  }

  /**
   * Coerce attendance status value
   * 'P', 'PRESENT', '1', 'YES' -> PRESENT
   * 'A', 'ABSENT', '0', 'NO' -> ABSENT
   * Otherwise -> NA
   */
  static coerceStatus(raw?: string): 'PRESENT' | 'ABSENT' | 'NA' {
    if (!raw) return 'NA';
    const clean = String(raw).trim().toLowerCase();
    
    if (['present', 'p', '1', 'yes'].includes(clean)) {
      return 'PRESENT';
    }
    if (['absent', 'a', '0', 'no'].includes(clean)) {
      return 'ABSENT';
    }
    return 'NA';
  }

  /**
   * Map raw row array to structured attendance entity
   */
  static mapRowToEntity(row: string[], config: RowMappingConfig, sourceRowHash: string): ParsedAttendanceRow | null {
    if (!row || row.length === 0) return null;

    const studentRollNo = this.coerceRollNo(row[config.rollNoCol]);
    const lectureDate = this.coerceDate(row[config.dateCol]);
    const rawStatus = row[config.statusCol];

    if (!studentRollNo || !lectureDate || !rawStatus) {
      return null;
    }

    const status = this.coerceStatus(rawStatus);

    return {
      studentRollNo,
      lectureDate,
      status,
      faculty: config.facultyCol !== undefined && row[config.facultyCol] ? String(row[config.facultyCol]).trim() : undefined,
      remarks: config.remarksCol !== undefined && row[config.remarksCol] ? String(row[config.remarksCol]).trim() : undefined,
      sourceRowHash
    };
  }
}

File: backend/src/main/java/com/vesit/openattend/service/sync/WorksheetMapper.java
package com.vesit.openattend.service.sync;

import com.vesit.openattend.entity.enums.AttendanceStatus;
import lombok.extern.slf4j.Slf4j;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@Slf4j
public class WorksheetMapper {

    private static final List<DateTimeFormatter> DATE_FORMATTERS = List.of(
            DateTimeFormatter.ofPattern("yyyy-MM-dd"),
            DateTimeFormatter.ofPattern("dd/MM/yyyy"),
            DateTimeFormatter.ofPattern("d/M/yyyy"),
            DateTimeFormatter.ofPattern("dd-MM-yyyy"),
            DateTimeFormatter.ofPattern("d-M-yyyy"),
            DateTimeFormatter.ofPattern("MM/dd/yyyy"),
            DateTimeFormatter.ofPattern("yyyy/MM/dd")
    );

    public static LocalDate parseDate(String rawDate) {
        if (rawDate == null || rawDate.trim().isEmpty()) {
            return null;
        }
        String clean = rawDate.trim();

        // Support Excel / Google Sheets numeric serial dates (e.g., 46237 or 46237.0)
        try {
            double serial = Double.parseDouble(clean);
            if (serial > 30000 && serial < 80000) {
                return LocalDate.of(1899, 12, 30).plusDays((long) serial);
            }
        } catch (NumberFormatException ignored) {
        }

        for (DateTimeFormatter formatter : DATE_FORMATTERS) {
            try {
                return LocalDate.parse(clean, formatter);
            } catch (DateTimeParseException ignored) {
            }
        }
        return null;
    }

    public static AttendanceStatus parseStatus(String rawStatus) {
        if (rawStatus == null || rawStatus.trim().isEmpty()) {
            return AttendanceStatus.NA;
        }
        String clean = rawStatus.trim().toUpperCase();
        if (clean.equals("PRESENT") || clean.equals("P") || clean.equals("1") || clean.equals("TRUE") || clean.equals("YES")) {
            return AttendanceStatus.PRESENT;
        }
        if (clean.equals("ABSENT") || clean.equals("A") || clean.equals("0") || clean.equals("FALSE") || clean.equals("NO")) {
            return AttendanceStatus.ABSENT;
        }
        return AttendanceStatus.NA;
    }

    public static ParsedAttendanceRow mapRowToEntity(List<Object> row, RowMappingConfig config, String sourceRowHash) {
        if (row == null || row.isEmpty() || config == null || config.getColumnRoles() == null) {
            return null;
        }

        Map<String, String> roles = config.getColumnRoles();
        int dateIdx = RowMappingConfig.columnLetterToIndex(roles.get("date"));
        int rollNoIdx = RowMappingConfig.columnLetterToIndex(roles.get("rollNo"));
        int statusIdx = RowMappingConfig.columnLetterToIndex(roles.get("status"));
        int facultyIdx = RowMappingConfig.columnLetterToIndex(roles.get("faculty"));
        int remarksIdx = RowMappingConfig.columnLetterToIndex(roles.get("remarks"));

        if (rollNoIdx < 0 || rollNoIdx >= row.size() || dateIdx < 0 || dateIdx >= row.size()) {
            return null;
        }

        String rawRollNo = row.get(rollNoIdx) != null ? row.get(rollNoIdx).toString().trim().toUpperCase() : null;
        String rawDate = row.get(dateIdx) != null ? row.get(dateIdx).toString().trim() : null;

        if (rawRollNo == null || rawRollNo.isEmpty() || rawDate == null || rawDate.isEmpty()) {
            return null;
        }

        LocalDate parsedDate = parseDate(rawDate);
        if (parsedDate == null) {
            return null;
        }

        String rawStatus = statusIdx >= 0 && statusIdx < row.size() && row.get(statusIdx) != null
                ? row.get(statusIdx).toString().trim()
                : null;
        AttendanceStatus status = parseStatus(rawStatus);

        String faculty = facultyIdx >= 0 && facultyIdx < row.size() && row.get(facultyIdx) != null
                ? row.get(facultyIdx).toString().trim()
                : null;

        String remarks = remarksIdx >= 0 && remarksIdx < row.size() && row.get(remarksIdx) != null
                ? row.get(remarksIdx).toString().trim()
                : null;

        return ParsedAttendanceRow.builder()
                .studentRollNo(rawRollNo)
                .lectureDate(parsedDate)
                .sessionIndex(0)
                .status(status)
                .faculty(faculty)
                .remarks(remarks)
                .sourceRowHash(sourceRowHash)
                .build();
    }

    /**
     * Unpivots VESIT professor matrix attendance grids (dates across columns, students in rows)
     */
    public static List<ParsedAttendanceRow> mapMatrixGridSheet(
            List<List<Object>> rows,
            int dateRowIndex,
            int startDataRowIndex,
            int rollNoColIndex,
            int startDateColIndex
    ) {
        List<ParsedAttendanceRow> results = new ArrayList<>();
        if (rows == null || rows.size() <= Math.max(dateRowIndex, startDataRowIndex)) {
            return results;
        }

        List<Object> dateRow = rows.get(dateRowIndex);
        List<LocalDate> columnDates = new ArrayList<>();
        for (int c = 0; c < dateRow.size(); c++) {
            if (c >= startDateColIndex && dateRow.get(c) != null) {
                LocalDate d = parseDate(dateRow.get(c).toString());
                columnDates.add(d);
            } else {
                columnDates.add(null);
            }
        }

        for (int r = startDataRowIndex; r < rows.size(); r++) {
            List<Object> row = rows.get(r);
            if (row == null || row.size() <= rollNoColIndex || row.get(rollNoColIndex) == null) {
                continue;
            }

            String roll = row.get(rollNoColIndex).toString().trim();
            if (roll.isEmpty()) continue;

            for (int c = startDateColIndex; c < Math.min(row.size(), columnDates.size()); c++) {
                LocalDate lectureDate = columnDates.get(c);
                if (lectureDate == null) continue;

                Object val = row.get(c);
                if (val == null) continue;
                String rawVal = val.toString().trim();
                if (rawVal.isEmpty()) continue;

                AttendanceStatus status = parseStatus(rawVal);
                if (status != AttendanceStatus.NA) {
                    results.add(ParsedAttendanceRow.builder()
                            .studentRollNo(roll)
                            .lectureDate(lectureDate)
                            .sessionIndex(0)
                            .status(status)
                            .build());
                }
            }
        }

        return results;
    }
}
